tests/Microsoft.Identity.Test.Integration.net45/HeadlessTests/ConfidentialClientIntegrationTests.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Security;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client;
+using Microsoft.Identity.Test.Integration.Infrastructure;
+using Microsoft.Identity.Test.LabInfrastructure;
+using Microsoft.Identity.Test.Unit;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Microsoft.Identity.Test.Integration.net45.HeadlessTests
+{
+    [TestClass]
+    public class ConfidentialClientIntegrationTests
+    {
+        private static readonly string[] s_scopes = { "User.Read" };
+        private static readonly string[] s_oboServiceScope = { "api://23c64cd8-21e4-41dd-9756-ab9e2c23f58c/access_as_user" };
+
+        [ClassInitialize]
+        public static void ClassInitialize(TestContext context)
+        {
+            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;
+        }
+
+        [TestMethod]
+        public async Task ClientSecretAuthenticationAsync()
+        {
+            var keyvault = new KeyVaultSecretsProvider();
+            var secret = keyvault.GetSecret(MsalTestConstants.MsalCCAKeyVaultUri).Value;
+            //TODO: acquire scenario specific client ids from the lab resonse
+            var confidentialClientID = "16dab2ba-145d-4b1b-8569-bf4b9aed4dc8";
+            var confidentialClientAuthority = "https://login.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47";
+            string[] scope = { "https://vault.azure.net/.default" };
+
+            var confidentialApp = ConfidentialClientApplicationBuilder
+                .Create(confidentialClientID)
+                .WithAuthority(new Uri(confidentialClientAuthority), true)
+                .WithClientSecret(secret)
+                .Build();
+
+            var authResult = await confidentialApp.AcquireTokenForClient(scope)
+                .ExecuteAsync(CancellationToken.None)
+                .ConfigureAwait(false);
+
+            MsalAssert.AssertAuthResult(authResult);
+        }
+
+        [TestMethod]
+        public async Task WebAPIAccessingGraphOnBehalfOfTestAsync()
+        {
+            var keyvault = new KeyVaultSecretsProvider();
+            var secret = keyvault.GetSecret(MsalTestConstants.MsalOBOKeyVaultUri).Value;
+            var labResponse = LabUserHelper.GetSpecificUser("IDLAB@msidlab4.onmicrosoft.com");
+            var user = labResponse.User;
+            //TODO: acquire scenario specific client ids from the lab resonse
+            var publicClientID = "be9b0186-7dfd-448a-a944-f771029105bf";
```

Comment:
I believe you have the public client id in labResponse.AppId?

Reply:
We should grab id's from the lab (something is in the lab response). If we have app id's that are not part of the lab, perhaps they should be ensuring we have only one place for app registrations?

---
In reply to: [279081003](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1118#discussion_r279081003) [](ancestors = 279081003)

Reply:
@bgavrilMS This client id is for an OBO app that is not the default app sent in the lab response.  The lab typically will create specific client ids for specific scenarios such as this. I will see if we can get additional client ids returned in the lab response instead of just the default.